firestore/bulkwriter.go

```diff
@@ -0,0 +1,332 @@
+	// IsOpen shows whether this BulkWriter is open or closed
+	IsOpen bool
+}
+
+// BulkWriterJob provides read-only access to the results of a BulkWriter write attempt.
+type BulkWriterJob struct {
+	op bulkWriterOperation // the operation this BulkWriterJob encapsulates
+}
+
+// Results gets the results of the BulkWriter write attempt
+// Attempting to access the results before the results have been received blocks
+// subsequent calls (until the result is received).
+func (j *BulkWriterJob) Results() (*WriteResult, error) {
+	return processResults(j.op)
+}
+
+// A BulkWriter allows multiple document writes in parallel. The BulkWriter
+// submits document writes in maximum batches of 20 writes per request. Each
+// request can contain many different document writes: create, delete, update,
+// and set are all supported. Only one operation per document is allowed.
+// Each call to Create, Update, Set, and Delete can return a value and error.
+type BulkWriter struct {
+	database        string             // the database as resource name: projects/[PROJECT]/databases/[DATABASE]
+	start           time.Time          // when this BulkWriter was started; used to calculate qps and rate increases
+	vc              *vkit.Client       // internal client
+	isOpen          bool               // flag that the BulkWriter is closed
+	maxOpsPerSecond int                // number of requests that can be sent per second
+	docUpdatePaths  map[string]bool    // document paths with corresponding writes in the queue
+	cancel          context.CancelFunc // context to send cancel message
+	bundler         *bundler.Bundler   // handle bundling up writes to Firestore
+	retryBundler    *bundler.Bundler   // handle bundling up retries of writes to Firestore
+	ctx             context.Context    // context for canceling BulkWriter operations
+	rateLock        sync.Mutex         // guards increases to rate of open sends; used in increaseRate()
+}
+
+// newBulkWriter creates a new instance of the BulkWriter. This
+// version of BulkWriter is intended to be used within go routines by the
+// callers.
+func newBulkWriter(ctx context.Context, c *Client, database string) *BulkWriter {
+	ctx, cancel := context.WithCancel(withResourceHeader(ctx, c.path()))
+
+	var rateLock sync.Mutex
+
+	bw := &BulkWriter{
+		database:        database,
+		start:           time.Now(),
+		vc:              c.c,
+		isOpen:          true,
+		maxOpsPerSecond: defaultStartingMaximumOpsPerSecond,
+		docUpdatePaths:  make(map[string]bool),
+		cancel:          cancel,
+		bundler:         &bundler.Bundler{},
+		retryBundler:    &bundler.Bundler{},
+		ctx:             ctx,
+		rateLock:        rateLock,
+	}
+
+	bw.bundler = bundler.NewBundler(bulkWriterOperation{}, bw.send)
+	bw.bundler.HandlerLimit = bw.maxOpsPerSecond
+	bw.bundler.BundleCountThreshold = maxBatchSize
+	bw.bundler.BundleByteLimit = 0 // unlimited size
+
+	bw.retryBundler = bundler.NewBundler(bulkWriterOperation{}, bw.send)
```

Comment:
Have not looked to see how or why this is used yet, but I wonder if the calls themselves should be retried and not have a seperate queue for this. May be uneeded complexity.

Reply:
Per our video chat, I'll check with the Firestore team about the reasoning here.